File: check-redirects.py
import os
import re
import csv
import time
import urllib.request
import urllib.parse
from concurrent.futures import ThreadPoolExecutor, as_completed

# Configuration
USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
CONCURRENCY = 15  # Process 15 requests at the same time

def extract_facebook_username(url):
    if not url:
        return None
    try:
        parsed = urllib.parse.urlparse(url)
        path = parsed.path.strip("/")
        
        if "profile.php" in path:
            queries = urllib.parse.parse_qs(parsed.query)
            if "id" in queries:
                return f"profile.php?id={queries['id'][0]}"
        
        parts = path.split("/")
        if parts:
            if parts[0] in ["pages", "people", "groups"]:
                if len(parts) > 1:
                    return parts[1]
            return parts[0]
    except Exception:
        pass
    return None

def check_facebook_url(url):
    req = urllib.request.Request(
        url, 
        headers={
            "User-Agent": USER_AGENT,
            "Accept-Language": "en-US,en;q=0.9",
        }
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            final_url = response.geturl()
            html = response.read().decode('utf-8', errors='ignore')
            
            canonical_match = re.search(r'<link\s+rel=["\']canonical["\']\s+href=["\']([^"\']+)["\']', html, re.IGNORECASE)
            if canonical_match:
                canonical_url = canonical_match.group(1)
                if "facebook.com" in canonical_url:
                    return canonical_url
            
            og_url_match = re.search(r'<meta\s+property=["\']og:url["\']\s+content=["\']([^"\']+)["\']', html, re.IGNORECASE)
            if og_url_match:
                og_url = og_url_match.group(1)
                if "facebook.com" in og_url:
                    return og_url
            
            return final_url
    except Exception:
        return None

def process_row(row, name_col, url_col):
    name = row.get(name_col, "Unknown").strip()
    url = row.get(url_col, "").strip()
    
    if not url or "facebook.com" not in url:
        return {"status": "SKIP", "name": name, "reason": "No valid Facebook URL found."}
        
    old_username = extract_facebook_username(url)
    if not old_username:
        return {"status": "SKIP", "name": name, "reason": "Could not parse username."}
        
    destination_url = check_facebook_url(url)
    if not destination_url:
        return {"status": "FAILED", "name": name, "reason": "Page unreachable."}
        
    new_username = extract_facebook_username(destination_url)
    
    if new_username and old_username.lower() != new_username.lower():
        return {
            "status": "CHANGED",
            "name": name,
            "old_url": url,
            "new_url": destination_url,
            "old_username": old_username,
            "new_username": new_username
        }
        
    return {"status": "OK", "name": name, "destination_url": destination_url}

def main():
    print("=" * 70)
    print("      FACEBOOK PAGE URL REDIRECTS CHECKER (PYTHON - BLAZING FAST)")
    print("=" * 70)
    
    files = [f for f in os.listdir(".") if f.endswith(".csv")]
    
    csv_file = None
    if len(files) == 1:
        csv_file = files[0]
    elif len(files) > 1:
        print("Multiple CSV files found in the current directory:")
        for idx, f in enumerate(files):
            print(f"  {idx + 1}. {f}")
        choice = input("Select the CSV file number: ").strip()
        try:
            csv_file = files[int(choice) - 1]
        except Exception:
            csv_file = choice
    else:
        csv_file = input("Please enter the name of your CSV file: ").strip()
        
    if not csv_file or not os.path.exists(csv_file):
        print(f"\n❌ Error: File '{csv_file}' not found.")
        return
        
    print(f"\nReading: '{csv_file}'...")
    
    rows = []
    with open(csv_file, mode='r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        fieldnames = reader.fieldnames
        for r in reader:
            rows.append(r)
            
    if not rows:
        print("❌ Error: No rows found or CSV is empty.")
        return
        
    name_col = next((c for c in fieldnames if "name" in c.lower()), fieldnames[0])
    url_col = next((c for c in fieldnames if "url" in c.lower() or "facebook" in c.lower() or "link" in c.lower()), fieldnames[1] if len(fieldnames) > 1 else fieldnames[0])
    
    print(f"Using column '{name_col}' for Page Names")
    print(f"Using column '{url_col}' for Page URLs")
    print(f"Scanning {len(rows)} pages with {CONCURRENCY} parallel threads...\n")
    
    print("-" * 70)
    print(f"{'PAGE NAME':<30} | {'STATUS':<10} | {'DETAILS'}")
    print("-" * 70)
    
    changes = []
    
    # Run requests concurrently using standard library ThreadPoolExecutor
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as executor:
        futures = {executor.submit(process_row, row, name_col, url_col): row for row in rows}
        
        for future in as_completed(futures):
            res = future.result()
            name = res["name"]
            
            if res["status"] == "SKIP":
                print(f"{name[:30]:<30} | \033[93mSKIP\033[0m       | {res['reason']}")
            elif res["status"] == "FAILED":
                print(f"{name[:30]:<30} | \033[91mFAILED\033[0m     | {res['reason']}")
            elif res["status"] == "OK":
                print(f"{name[:30]:<30} | OK         | Active URL: {res['destination_url']}")
            elif res["status"] == "CHANGED":
                print(f"\033[92m{name[:30]:<30}\033[0m | \033[92mCHANGED\033[0m    | {res['old_username']} ➜ {res['new_username']}")
                print(f"  ↳ Old: {res['old_url']}")
                print(f"  ↳ New: {res['new_url']}")
                changes.append({
                    "Page Name": name,
                    "Old URL": res["old_url"],
                    "New URL": res["new_url"],
                    "Old Username": res["old_username"],
                    "New Username": res["new_username"]
                })
                
    print("-" * 70)
    print(f"\nScan complete! Checked {len(rows)} pages.")
    print(f"Detected {len(changes)} changed Facebook URLs.")
    
    if changes:
        report_file = "url_changes_report.csv"
        with open(report_file, mode='w', encoding='utf-8', newline='') as f:
            writer = csv.DictWriter(f, fieldnames=["Page Name", "Old URL", "New URL", "Old Username", "New Username"])
            writer.writeheader()
            writer.writerows(changes)
        print(f"📄 Saved list of changes to: '\033[92m{report_file}\033[0m'")
    else:
        print("No changes found!")

if __name__ == "__main__":
    main()

File: check-redirects.js
import fs from 'fs';
import path from 'path';
import readline from 'readline';

const USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
const REQUEST_DELAY = 2000; // 2 seconds delay to avoid Facebook rate limits

function extractFacebookUsername(url) {
  if (!url) return null;
  try {
    const parsed = new URL(url);
    let pathname = parsed.pathname.replace(/^\/|\/$/g, ''); // strip leading/trailing slashes
    
    if (pathname.includes('profile.php')) {
      const id = parsed.searchParams.get('id');
      if (id) return `profile.php?id=${id}`;
    }
    
    const parts = pathname.split('/');
    if (parts.length > 0) {
      if (['pages', 'people', 'groups'].includes(parts[0])) {
        if (parts.length > 1) return parts[1];
      }
      return parts[0];
    }
  } catch (e) {}
  return null;
}

async function checkFacebookUrl(url) {
  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': USER_AGENT,
        'Accept-Language': 'en-US,en;q=0.9',
      },
      redirect: 'follow'
    });
    
    const finalUrl = response.url;
    const html = await response.text();
    
    // 1. Search for canonical link in HTML
    const canonicalMatch = html.match(/<link\s+rel=["']canonical["']\s+href=["']([^"']+)["']/i);
    if (canonicalMatch && canonicalMatch[1]) {
      const canonicalUrl = canonicalMatch[1];
      if (canonicalUrl.includes('facebook.com')) {
        return canonicalUrl;
      }
    }
    
    // 2. Search for og:url meta tag
    const ogMatch = html.match(/<meta\s+property=["']og:url["']\s+content=["']([^"']+)["']/i);
    if (ogMatch && ogMatch[1]) {
      const ogUrl = ogMatch[1];
      if (ogUrl.includes('facebook.com')) {
        return ogUrl;
      }
    }
    
    return finalUrl;
  } catch (e) {
    return null;
  }
}

// Minimal CSV parser
function parseCSV(content) {
  const lines = content.split(/\r?\n/);
  if (lines.length === 0) return { fieldnames: [], rows: [] };
  
  // Simple CSV line splitter (handles quotes roughly but works perfectly for standard Google sheets CSV)
  const splitCSVLine = (line) => {
    const result = [];
    let current = '';
    let inQuotes = false;
    for (let i = 0; i < line.length; i++) {
      const char = line[i];
      if (char === '"') {
        inQuotes = !inQuotes;
      } else if (char === ',' && !inQuotes) {
        result.push(current.trim());
        current = '';
      } else {
        current += char;
      }
    }
    result.push(current.trim());
    return result.map(v => v.replace(/^"|"$/g, '').trim());
  };
  
  const fieldnames = splitCSVLine(lines[0]);
  const rows = [];
  
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i].trim()) continue;
    const values = splitCSVLine(lines[i]);
    const row = {};
    fieldnames.forEach((name, idx) => {
      row[name] = values[idx] || '';
    });
    rows.push(row);
  }
  
  return { fieldnames, rows };
}

// Minimal CSV writer
function writeCSV(filepath, fieldnames, rows) {
  const headerLine = fieldnames.map(f => `"${f.replace(/"/g, '""')}"`).join(',');
  const rowLines = rows.map(row => 
    fieldnames.map(f => {
      const val = row[f] || '';
      return `"${val.replace(/"/g, '""')}"`;
    }).join(',')
  );
  fs.writeFileSync(filepath, [headerLine, ...rowLines].join('\n'), 'utf8');
}

async function main() {
  console.log("======================================================================");
  console.log("      FACEBOOK PAGE URL REDIRECTS CHECKER (NODE.JS - ZERO INSTALLS)");
  console.log("======================================================================\n");
  
  const files = fs.readdirSync('.').filter(f => f.endsWith('.csv'));
  let csvFile = null;
  
  if (files.length === 1) {
    csvFile = files[0];
  } else if (files.length > 1) {
    console.log("Multiple CSV files found in the current directory:");
    files.forEach((f, idx) => console.log(`  ${idx + 1}. ${f}`));
    
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const answer = await new Promise(resolve => rl.question('\nSelect the CSV file number: ', resolve));
    rl.close();
    
    const idx = parseInt(answer.trim(), 10) - 1;
    csvFile = files[idx] || files[0];
  } else {
    console.log("❌ Error: No CSV file found in the current directory.");
    console.log("Tip: Export your Google Sheet as a CSV file (File -> Download -> Comma-separated values .csv)");
    console.log("and place it inside this folder before running the script.");
    return;
  }
  
  console.log(`Reading: '${csvFile}'...`);
  const content = fs.readFileSync(csvFile, 'utf8');
  const { fieldnames, rows } = parseCSV(content);
  
  if (rows.length === 0) {
    console.log("❌ Error: No rows found or CSV is empty.");
    return;
  }
  
  // Detect columns
  let nameCol = fieldnames.find(f => f.toLowerCase().includes('name'));
  let urlCol = fieldnames.find(f => f.toLowerCase().includes('url') || f.toLowerCase().includes('facebook') || f.toLowerCase().includes('link'));
  
  if (!nameCol) nameCol = fieldnames[0];
  if (!urlCol) urlCol = fieldnames[1] || fieldnames[0];
  
  console.log(`Using column '${nameCol}' for Page Names`);
  console.log(`Using column '${urlCol}' for Page URLs`);
  console.log(`Scanning ${rows.length} pages...\n`);
  
  console.log("----------------------------------------------------------------------");
  console.log(`${"PAGE NAME".padEnd(30)} | ${"STATUS".padEnd(10)} | DETAILS`);
  console.log("----------------------------------------------------------------------");
  
  const changes = [];
  
  for (const row of rows) {
    const name = row[nameCol] || 'Unknown';
    const url = row[urlCol] || '';
    
    if (!url || !url.includes('facebook.com')) {
      console.log(`\x1b[33m${name.substring(0, 30).padEnd(30)}\x1b[0m | SKIP       | No valid Facebook URL found.`);
      continue;
    }
    
    const oldUsername = extractFacebookUsername(url);
    if (!oldUsername) {
      console.log(`\x1b[33m${name.substring(0, 30).padEnd(30)}\x1b[0m | SKIP       | Could not parse username from URL.`);
      continue;
    }
    
    // Add delay between requests
    await new Promise(resolve => setTimeout(resolve, REQUEST_DELAY));
    
    const destinationUrl = await checkFacebookUrl(url);
    if (!destinationUrl) {
      console.log(`\x1b[31m${name.substring(0, 30).padEnd(30)}\x1b[0m | FAILED     | Page unreachable or blocked.`);
      continue;
    }
    
    const newUsername = extractFacebookUsername(destinationUrl);
    
    if (newUsername && oldUsername.toLowerCase() !== newUsername.toLowerCase()) {
      console.log(`\x1b[32m${name.substring(0, 30).padEnd(30)}\x1b[0m | CHANGED    | ${oldUsername} ➜ ${newUsername}`);
      console.log(`  ↳ Old: ${url}`);
      console.log(`  ↳ New: ${destinationUrl}`);
      changes.push({
        "Page Name": name,
        "Old URL": url,
        "New URL": destinationUrl,
        "Old Username": oldUsername,
        "New Username": newUsername
      });
    } else {
      console.log(`${name.substring(0, 30).padEnd(30)} | OK         | URL is active and unchanged.`);
    }
  }
  
  console.log("----------------------------------------------------------------------");
  console.log(`\nScan complete! Checked ${rows.length} pages.`);
  console.log(`Detected ${changes.length} changed Facebook URLs.`);
  
  if (changes.length > 0) {
    const reportFile = "url_changes_report.csv";
    writeCSV(reportFile, ["Page Name", "Old URL", "New URL", "Old Username", "New Username"], changes);
    console.log(`📄 Saved list of changes to: \x1b[32m${reportFile}\x1b[0m`);
  } else {
    console.log("No changes found!");
  }
}

main().catch(console.error);
